BEST_PRACTICES.md

```diff
@@ -0,0 +1,84 @@
+real nodes for your parallelized work.
+* Nested `parallel` blocks can lead to swamping your available
+executors, as each execution of the first `parallel` block calls
+multiple executions of the second `parallel` block, and so on. In
+general, think carefully about your parallelism and your available executors
+when using `parallel`.
+* The [Parallel Test Executor
+plugin](https://github.com/jenkinsci/parallel-test-executor-plugin) is
+awesome and can be immensely helpful both for distributing your test
+execution and for throttling your parallelism, since you define how
+many "buckets" your tests get divided into.
+* Don’t put `stage`s in `parallel` blocks - that just goes *weird*,
+breaking a lot of logic in the Stage View and elsewhere. Save yourself
+the pain - don't do it!
+
+# `Jenkinsfile`s and Multibranch
+* Use: `checkout scm` to automatically checkout current revision of branch
+* Use: `$env.BRANCH_NAME` variable if you have logical difference in
+your flow between branches, i.e. to distinguish different behavior for
+production-ready branches versus sandbox or pull request branches.
+* For `Jenkinsfile`s, make sure to put `#!groovy` at the top of the
+file so that IDEs, GitHub diffs, etc properly detect the language and
+do syntax highlighting for you.
+
+# Groovy gotchas
+* Don’t have the Groovy interpreter making blocking i/o calls, i.e.,
+`HTTPClient` and the like - these can cause real problems with
+resumability, and also require a lot of explicit whitelisting of
+methods in the Script Security plugin, which is not ideal.
+* Beware `for (Foo f: foos)` loops and Groovy closure-style operators
+like `.each` and the like. They will not work right in normal Workflow
+script contexts where Workflow steps are involved directly.
+* If you need to do that kind of thing to make your scripting make
+sense, do it in methods annotated with `@NonCPS`. These methods will
+not be CPS-transformed, so can do some things that just can't be done
+in the serializable/resumable CPS context.
+* Don’t use the Groovy scripting in place of shell scripting - work
+coming for the ability to run a Groovy step on the node as with the
+normal Groovy plugin build step, but until then, shell out, even if
+it’s just to do `sh 'groovy foo.groovy'`.
+* If you really need `Map`s in your normal Workflow scripts, consider
+using arrays of `[key,value]` instead - arrays have serializable
+`Iterator`s and so can be used more normally than `Map`s can.
+
+# Workflow script development tips
+* When developing new flows, you can often iterate faster with an
+inline workflow (not from SCM) but using the 'load' operation to load
+common utility methods from common workflows, and then as you finish
+out methods commit them to the utility flows.  This lets you strike a
+balance between having traceability on commits and being able to move
+fast. NOTE: this isn't possible with Multibranch Workflows, since
+those *have* to pull their script from SCM completely.
```

Comment:
You can still create a workflow to clone a specific branch and run flow steps though... probably best for fancy work where you'll have to do a lot of debugging around small issues (which would yield dozens of single-line commits).